src/main/java/com/gw2auth/oauth2/server/service/gw2account/verification/Gw2AccountVerificationServiceImpl.java

```diff
@@ -276,33 +277,38 @@ public boolean verify(AccountService.LoggingContext logging, Gw2AccountVerificat
         return challenge.verify(challenge.readState(rawState), gw2ApiToken);
     }
 
     @Scheduled(fixedRate = 60L, timeUnit = TimeUnit.SECONDS)
+    @Transactional
     public void tryVerifyAllPending() {
         final Instant now = this.clock.instant();
         final List<Gw2AccountVerificationChallengePendingEntity> entities = this.gw2AccountVerificationChallengePendingRepository.findAll();
+        if (entities.isEmpty()) {
+            return;
+        }
 
-        if (!entities.isEmpty()) {
-            LOG.info("processing {} pending challenges", entities.size());
+        LOG.info("processing {} pending challenges", entities.size());
 
-            for (Gw2AccountVerificationChallengePendingEntity entity : entities) {
-                final UUID accountId = entity.accountId();
-                final UUID gw2AccountId = entity.gw2AccountId();
+        for (Gw2AccountVerificationChallengePendingEntity entity : entities) {
+            final UUID accountId = entity.accountId();
+            final UUID gw2AccountId = entity.gw2AccountId();
+            final VerificationChallenge<?> challenge = this.challengesById.get(entity.challengeId());
 
-                try (AccountService.LoggingContext logging = this.accountService.log(accountId, Map.of("type", "gw2.verification.attempt", "gw2_account_id", gw2AccountId, "challenge_id", entity.challengeId()))) {
-                    logging.log("Trying to verify GW2 account");
+            try (AccountService.LoggingContext logging = this.accountService.log(accountId, Map.of("type", "gw2.verification.attempt", "gw2_account_id", gw2AccountId, "challenge_id", entity.challengeId()))) {
+                logging.log("Trying to verify GW2 account");
 
+                if (challenge == null) {
+                    logging.log("Challenge was no longer found; might be deactivated; trying again on next schedule");
+                } else if (now.isAfter(entity.timeoutTime())) {
+                    this.gw2AccountVerificationChallengePendingRepository.deleteByAccountIdAndGw2AccountId(entity.accountId(), entity.gw2AccountId());
+                    logging.log("GW2 account failed to verify within the allowed period.");
+                    recordMetric(entity, false);
+                } else {
                     try {
-                        if (now.isAfter(entity.timeoutTime())) {
-                            this.gw2AccountVerificationChallengePendingRepository.deleteByAccountIdAndGw2AccountId(entity.accountId(), entity.gw2AccountId());
-                            logging.log("GW2 account failed to verify within the allowed period.");
-                            recordMetric(entity, false);
-                        } else {
-                            verify(logging, entity);
-                        }
+                        verifyInternal(logging, entity, challenge);
                     } catch (Exception e) {
```

Comment:
> During scheduled runs, a rollback should essentially never happen (unless there's a completely unhandled exception). The re-thrown IOException by verifyInternal is caught and thus not rolled back.

There won't be any rollback for unexpected exceptions during the `verifyInternal` call in scheduled runs because any `Exception` is caught here and not rethrown. If I understood you right, this is not what you intended to do here. Could you double-check this?

Reply:
youre right, this should only catch `IOException`.